llvm/include/llvm/Analysis/GenericDomTreeUpdaterImpl.h

```diff
@@ -0,0 +1,477 @@
+  return *static_cast<DerivedT *>(this);
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+const DerivedT &
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::derived() const {
+  return *static_cast<const DerivedT *>(this);
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    UpdateStrategy Strategy_)
+    : Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    DomTreeT &DT_, UpdateStrategy Strategy_)
+    : DT(&DT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    DomTreeT *DT_, UpdateStrategy Strategy_)
+    : DT(DT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    PostDomTreeT &PDT_, UpdateStrategy Strategy_)
+    : PDT(&PDT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    PostDomTreeT *PDT_, UpdateStrategy Strategy_)
+    : PDT(PDT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    DomTreeT &DT_, PostDomTreeT &PDT_, UpdateStrategy Strategy_)
+    : DT(&DT_), PDT(&PDT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::GenericDomTreeUpdater(
+    DomTreeT *DT_, PostDomTreeT *PDT_, UpdateStrategy Strategy_)
+    : DT(DT_), PDT(PDT_), Strategy(Strategy_) {}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+GenericDomTreeUpdater<DerivedT, DomTreeT,
+                      PostDomTreeT>::~GenericDomTreeUpdater() {
+  flush();
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+bool GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::isLazy() const {
+  return Strategy == UpdateStrategy::Lazy;
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+bool GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::isEager() const {
+  return Strategy == UpdateStrategy::Eager;
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+bool GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::hasDomTree()
+    const {
+  return DT != nullptr;
+}
+
+template <typename DerivedT, typename DomTreeT, typename PostDomTreeT>
+bool GenericDomTreeUpdater<DerivedT, DomTreeT, PostDomTreeT>::hasPostDomTree()
+    const {
+  return DT != nullptr;
+}
```

Comment:
It's fine (and preferable) to keep the definitions for all these tiny single-line methods in the header, assuming they work fine with forward-declarations, which I think they do. Basically the ones that were in the header prior to https://github.com/llvm/llvm-project/pull/96851 can also stay in the header now.